In [2]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

In [3]:
from pathlib import Path

In [4]:
PROJECT_ROOT = Path.cwd().parent
DATA_PATH = PROJECT_ROOT / "data" / "raw" / "WA_Fn-UseC_-HR-Employee-Attrition.csv"

print(DATA_PATH)

d:\PORTOFOLIO DATA ANALYST\hr-workforce-analytics\data\raw\WA_Fn-UseC_-HR-Employee-Attrition.csv


In [5]:
df = pd.read_csv(DATA_PATH)

In [6]:
df.head()

,Age,Attrition,BusinessTravel,DailyRate,Department,DistanceFromHome,Education,EducationField,EmployeeCount,EmployeeNumber,...,RelationshipSatisfaction,StandardHours,StockOptionLevel,TotalWorkingYears,TrainingTimesLastYear,WorkLifeBalance,YearsAtCompany,YearsInCurrentRole,YearsSinceLastPromotion,YearsWithCurrManager
0,41,Yes,Travel_Rarely,1102,Sales,1,2,Life Sciences,1,1,...,1,80,0,8,0,1,6,4,0,5
1,49,No,Travel_Frequently,279,Research & Development,8,1,Life Sciences,1,2,...,4,80,1,10,3,3,10,7,1,7
2,37,Yes,Travel_Rarely,1373,Research & Development,2,2,Other,1,4,...,2,80,0,7,3,3,0,0,0,0
3,33,No,Travel_Frequently,1392,Research & Development,3,4,Life Sciences,1,5,...,3,80,0,8,3,3,8,7,3,0
4,27,No,Travel_Rarely,591,Research & Development,2,1,Medical,1,7,...,4,80,1,6,3,3,2,2,2,2


In [7]:
df.tail()

,Age,Attrition,BusinessTravel,DailyRate,Department,DistanceFromHome,Education,EducationField,EmployeeCount,EmployeeNumber,...,RelationshipSatisfaction,StandardHours,StockOptionLevel,TotalWorkingYears,TrainingTimesLastYear,WorkLifeBalance,YearsAtCompany,YearsInCurrentRole,YearsSinceLastPromotion,YearsWithCurrManager
1465,36,No,Travel_Frequently,884,Research & Development,23,2,Medical,1,2061,...,3,80,1,17,3,3,5,2,0,3
1466,39,No,Travel_Rarely,613,Research & Development,6,1,Medical,1,2062,...,1,80,1,9,5,3,7,7,1,7
1467,27,No,Travel_Rarely,155,Research & Development,4,3,Life Sciences,1,2064,...,2,80,1,6,0,3,6,2,0,3
1468,49,No,Travel_Frequently,1023,Sales,2,3,Medical,1,2065,...,4,80,0,17,3,2,9,6,0,8
1469,34,No,Travel_Rarely,628,Research & Development,8,3,Medical,1,2068,...,1,80,0,6,3,4,4,3,1,2


In [8]:
print(f"Rows    : {df.shape[0]:,}")
print(f"Columns : {df.shape[1]:,}")

Rows    : 1,470
Columns : 35


In [9]:
df.shape

(1470, 35)

In [10]:
df.columns.tolist()

['Age',
 'Attrition',
 'BusinessTravel',
 'DailyRate',
 'Department',
 'DistanceFromHome',
 'Education',
 'EducationField',
 'EmployeeCount',
 'EmployeeNumber',
 'EnvironmentSatisfaction',
 'Gender',
 'HourlyRate',
 'JobInvolvement',
 'JobLevel',
 'JobRole',
 'JobSatisfaction',
 'MaritalStatus',
 'MonthlyIncome',
 'MonthlyRate',
 'NumCompaniesWorked',
 'Over18',
 'OverTime',
 'PercentSalaryHike',
 'PerformanceRating',
 'RelationshipSatisfaction',
 'StandardHours',
 'StockOptionLevel',
 'TotalWorkingYears',
 'TrainingTimesLastYear',
 'WorkLifeBalance',
 'YearsAtCompany',
 'YearsInCurrentRole',
 'YearsSinceLastPromotion',
 'YearsWithCurrManager']

In [11]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 1470 entries, 0 to 1469
Data columns (total 35 columns):
 #   Column                    Non-Null Count  Dtype
---  ------                    --------------  -----
 0   Age                       1470 non-null   int64
 1   Attrition                 1470 non-null   str  
 2   BusinessTravel            1470 non-null   str  
 3   DailyRate                 1470 non-null   int64
 4   Department                1470 non-null   str  
 5   DistanceFromHome          1470 non-null   int64
 6   Education                 1470 non-null   int64
 7   EducationField            1470 non-null   str  
 8   EmployeeCount             1470 non-null   int64
 9   EmployeeNumber            1470 non-null   int64
 10  EnvironmentSatisfaction   1470 non-null   int64
 11  Gender                    1470 non-null   str  
 12  HourlyRate                1470 non-null   int64
 13  JobInvolvement            1470 non-null   int64
 14  JobLevel                  1470 non-null   int64
 15

In [12]:
missing = (
    df.isna()
      .sum()
      .to_frame("missing_count")
)

missing["missing_percentage"] = (
    missing["missing_count"] / len(df) * 100
)

missing.sort_values(
    "missing_count",
    ascending=False
)

,missing_count,missing_percentage
Age,0,0.0
Attrition,0,0.0
BusinessTravel,0,0.0
DailyRate,0,0.0
Department,0,0.0
DistanceFromHome,0,0.0
Education,0,0.0
EducationField,0,0.0
EmployeeCount,0,0.0
EmployeeNumber,0,0.0


In [13]:
duplicate_count = df.duplicated().sum()

print(f"Duplicate rows: {duplicate_count:,}")

Duplicate rows: 0


In [14]:
unique_summary = pd.DataFrame({
    "column": df.columns,
    "unique_count": [df[col].nunique(dropna=False) for col in df.columns],
    "dtype": [df[col].dtype for col in df.columns]
})

unique_summary

,column,unique_count,dtype
0,Age,43,int64
1,Attrition,2,str
2,BusinessTravel,3,str
3,DailyRate,886,int64
4,Department,3,str
5,DistanceFromHome,29,int64
6,Education,5,int64
7,EducationField,6,str
8,EmployeeCount,1,int64
9,EmployeeNumber,1470,int64


In [15]:
df.describe().T

,count,mean,std,min,25%,50%,75%,max
Age,1470.0,36.923810,9.135373,18.0,30.00,36.0,43.00,60.0
DailyRate,1470.0,802.485714,403.509100,102.0,465.00,802.0,1157.00,1499.0
DistanceFromHome,1470.0,9.192517,8.106864,1.0,2.00,7.0,14.00,29.0
Education,1470.0,2.912925,1.024165,1.0,2.00,3.0,4.00,5.0
EmployeeCount,1470.0,1.000000,0.000000,1.0,1.00,1.0,1.00,1.0
EmployeeNumber,1470.0,1024.865306,602.024335,1.0,491.25,1020.5,1555.75,2068.0
EnvironmentSatisfaction,1470.0,2.721769,1.093082,1.0,2.00,3.0,4.00,4.0
HourlyRate,1470.0,65.891156,20.329428,30.0,48.00,66.0,83.75,100.0
JobInvolvement,1470.0,2.729932,0.711561,1.0,2.00,3.0,3.00,4.0
JobLevel,1470.0,2.063946,1.106940,1.0,1.00,2.0,3.00,5.0


In [19]:
categorical_columns = df.select_dtypes(
    include=["str"]
).columns.tolist()

categorical_columns

['Attrition',
 'BusinessTravel',
 'Department',
 'EducationField',
 'Gender',
 'JobRole',
 'MaritalStatus',
 'Over18',
 'OverTime']

In [17]:
for col in categorical_columns:
    print(f"\n--- {col} ---")
    print(df[col].value_counts(dropna=False).head(10))


--- Attrition ---
Attrition
No     1233
Yes     237
Name: count, dtype: int64

--- BusinessTravel ---
BusinessTravel
Travel_Rarely        1043
Travel_Frequently     277
Non-Travel            150
Name: count, dtype: int64

--- Department ---
Department
Research & Development    961
Sales                     446
Human Resources            63
Name: count, dtype: int64

--- EducationField ---
EducationField
Life Sciences       606
Medical             464
Marketing           159
Technical Degree    132
Other                82
Human Resources      27
Name: count, dtype: int64

--- Gender ---
Gender
Male      882
Female    588
Name: count, dtype: int64

--- JobRole ---
JobRole
Sales Executive              326
Research Scientist           292
Laboratory Technician        259
Manufacturing Director       145
Healthcare Representative    131
Manager                      102
Sales Representative          83
Research Director             80
Human Resources               52
Name: count, dtype: int

In [18]:
attrition_rate = (
    df["Attrition"].value_counts(normalize=True) * 100
)

attrition_rate

Attrition
No     83.877551
Yes    16.122449
Name: proportion, dtype: float64

In [20]:
constant_columns = [
    col for col in df.columns
    if df[col].nunique(dropna=False) <= 1
]

constant_columns

['EmployeeCount', 'Over18', 'StandardHours']

In [23]:
identifier_check = {
    "EmployeeNumber": {
        "row_count": len(df),
        "unique_count": df["EmployeeNumber"].nunique(),
        "is_unique": df["EmployeeNumber"].is_unique
    }
}

identifier_check

{'EmployeeNumber': {'row_count': 1470,
  'unique_count': 1470,
  'is_unique': True}}

In [24]:
total_employees = len(df)

attrition_count = (
    df["Attrition"]
    .eq("Yes")
    .sum()
)

attrition_rate = (
    attrition_count / total_employees * 100
)

print(f"Total Employees : {total_employees:,}")
print(f"Attrition       : {attrition_count:,}")
print(f"Attrition Rate  : {attrition_rate:.2f}%")

Total Employees : 1,470
Attrition       : 237
Attrition Rate  : 16.12%


In [25]:
quality_summary = pd.DataFrame({
    "Metric": [
        "Total Rows",
        "Total Columns",
        "Missing Values",
        "Duplicate Rows",
        "Constant Columns",
        "Unique Employee Numbers",
        "Attrition Count",
        "Attrition Rate"
    ],
    "Value": [
        len(df),
        df.shape[1],
        int(df.isna().sum().sum()),
        int(df.duplicated().sum()),
        len(constant_columns),
        int(df["EmployeeNumber"].nunique()),
        int(attrition_count),
        f"{attrition_rate:.2f}%"
    ]
})

quality_summary

,Metric,Value
0,Total Rows,1470
1,Total Columns,35
2,Missing Values,0
3,Duplicate Rows,0
4,Constant Columns,3
5,Unique Employee Numbers,1470
6,Attrition Count,237
7,Attrition Rate,16.12%


# Business Understanding

## Business Objective

Analyze employee attrition, workforce characteristics, job satisfaction,
compensation, workload, and career-related factors to identify workforce
segments and factors associated with employee turnover.

## Key Business Questions

### Workforce Overview
- What is the size and composition of the workforce?
- How are employees distributed across departments and job roles?

### Employee Attrition
- What is the overall attrition rate?
- Which departments and job roles have the highest attrition rates?
- How does attrition vary across demographic groups?

### Workload & Work-Life Balance
- Is overtime associated with higher attrition?
- How does work-life balance relate to attrition?
- Does distance from home relate to attrition?

### Compensation & Career
- How does attrition vary by job level and income?
- Is stock option level associated with attrition?
- Does shorter tenure correspond to higher attrition?

### Satisfaction & Performance
- How does job satisfaction relate to attrition?
- How does environment satisfaction relate to attrition?
- Is attrition different across performance ratings?

# Data Quality Assessment

## Dataset Overview

- Total rows: 1,470
- Total columns: 35
- Missing values: 0
- Duplicate rows: 0
- Unique employee numbers: 1,470
- Attrition cases: 237
- Overall attrition rate: 16.12%

## Data Quality Findings

### Missing Values
No missing values were identified across the dataset.

### Duplicate Records
No duplicate rows were identified.

### Identifier
`EmployeeNumber` contains 1,470 unique values across 1,470 employees
and can be used as the employee-level identifier.

### Constant Variables
The following variables contain only one unique value:

- `EmployeeCount`
- `Over18`
- `StandardHours`

These variables provide no analytical variation and are candidates
for removal during the data cleaning stage.

### Target Variable
`Attrition` contains two categories:

- No: 1,233 employees
- Yes: 237 employees

The overall attrition rate is 16.12%.

# Data Dictionary

The data dictionary describes the business meaning, analytical role,
and treatment of each variable in the HR workforce dataset.

## Data Dictionary Interpretation

The dataset contains 35 variables covering employee demographics,
employment characteristics, compensation, satisfaction, performance,
and career history.

Variables are classified into five analytical roles:

- Identifier
- Target
- Feature
- Ordinal
- Redundant

Three constant variables (`EmployeeCount`, `Over18`, and `StandardHours`)
are identified as redundant because they contain no variation across
employees and provide no analytical value.

In [26]:
data_dictionary = pd.DataFrame([
    ["Age", "int64", "Employee age", "Feature", "Keep"],
    ["Attrition", "string", "Whether the employee left the company", "Target", "Keep"],
    ["BusinessTravel", "string", "Frequency of business travel", "Feature", "Keep"],
    ["DailyRate", "int64", "Daily rate-related employee value", "Feature", "Keep"],
    ["Department", "string", "Employee department", "Feature", "Keep"],
    ["DistanceFromHome", "int64", "Distance between employee home and workplace", "Feature", "Keep"],
    ["Education", "int64", "Education level", "Ordinal", "Keep"],
    ["EducationField", "string", "Employee education field", "Feature", "Keep"],
    ["EmployeeCount", "int64", "Employee count indicator", "Redundant", "Remove"],
    ["EmployeeNumber", "int64", "Unique employee identifier", "Identifier", "Keep"],
    ["EnvironmentSatisfaction", "int64", "Employee satisfaction with work environment", "Ordinal", "Keep"],
    ["Gender", "string", "Employee gender", "Feature", "Keep"],
    ["HourlyRate", "int64", "Hourly rate-related employee value", "Feature", "Keep"],
    ["JobInvolvement", "int64", "Level of employee involvement in the job", "Ordinal", "Keep"],
    ["JobLevel", "int64", "Employee organizational/job level", "Ordinal", "Keep"],
    ["JobRole", "string", "Employee job role", "Feature", "Keep"],
    ["JobSatisfaction", "int64", "Employee job satisfaction level", "Ordinal", "Keep"],
    ["MaritalStatus", "string", "Employee marital status", "Feature", "Keep"],
    ["MonthlyIncome", "int64", "Employee monthly income", "Feature", "Keep"],
    ["MonthlyRate", "int64", "Monthly rate-related employee value", "Feature", "Keep"],
    ["NumCompaniesWorked", "int64", "Number of companies previously worked for", "Feature", "Keep"],
    ["Over18", "string", "Whether employee is over 18 years old", "Redundant", "Remove"],
    ["OverTime", "string", "Whether employee works overtime", "Feature", "Keep"],
    ["PercentSalaryHike", "int64", "Percentage increase in salary", "Feature", "Keep"],
    ["PerformanceRating", "int64", "Employee performance rating", "Ordinal", "Keep"],
    ["RelationshipSatisfaction", "int64", "Employee relationship satisfaction level", "Ordinal", "Keep"],
    ["StandardHours", "int64", "Standard working hours", "Redundant", "Remove"],
    ["StockOptionLevel", "int64", "Employee stock option level", "Ordinal", "Keep"],
    ["TotalWorkingYears", "int64", "Total years of work experience", "Feature", "Keep"],
    ["TrainingTimesLastYear", "int64", "Number of training sessions attended last year", "Feature", "Keep"],
    ["WorkLifeBalance", "int64", "Employee work-life balance rating", "Ordinal", "Keep"],
    ["YearsAtCompany", "int64", "Years employee has worked at the company", "Feature", "Keep"],
    ["YearsInCurrentRole", "int64", "Years employee has been in current role", "Feature", "Keep"],
    ["YearsSinceLastPromotion", "int64", "Years since employee's last promotion", "Feature", "Keep"],
    ["YearsWithCurrManager", "int64", "Years working with current manager", "Feature", "Keep"],
], columns=[
    "column",
    "data_type",
    "description",
    "analytical_role",
    "treatment"
])

data_dictionary

,column,data_type,description,analytical_role,treatment
0,Age,int64,Employee age,Feature,Keep
1,Attrition,string,Whether the employee left the company,Target,Keep
2,BusinessTravel,string,Frequency of business travel,Feature,Keep
3,DailyRate,int64,Daily rate-related employee value,Feature,Keep
4,Department,string,Employee department,Feature,Keep
5,DistanceFromHome,int64,Distance between employee home and workplace,Feature,Keep
6,Education,int64,Education level,Ordinal,Keep
7,EducationField,string,Employee education field,Feature,Keep
8,EmployeeCount,int64,Employee count indicator,Redundant,Remove
9,EmployeeNumber,int64,Unique employee identifier,Identifier,Keep


In [30]:
data_dictionary["unique_count"] = (
    data_dictionary["column"]
    .map(df.nunique(dropna=False))
)

data_dictionary = data_dictionary[
    [
        "column",
        "data_type",
        "unique_count",
        "analytical_role",
        "description",
        "treatment"
    ]
]

data_dictionary

,column,data_type,unique_count,analytical_role,description,treatment
0,Age,int64,43,Feature,Employee age,Keep
1,Attrition,string,2,Target,Whether the employee left the company,Keep
2,BusinessTravel,string,3,Feature,Frequency of business travel,Keep
3,DailyRate,int64,886,Feature,Daily rate-related employee value,Keep
4,Department,string,3,Feature,Employee department,Keep
5,DistanceFromHome,int64,29,Feature,Distance between employee home and workplace,Keep
6,Education,int64,5,Ordinal,Education level,Keep
7,EducationField,string,6,Feature,Employee education field,Keep
8,EmployeeCount,int64,1,Redundant,Employee count indicator,Remove
9,EmployeeNumber,int64,1470,Identifier,Unique employee identifier,Keep


# Range Validation & Outlier Assessment

Range validation is performed to identify values that may violate
expected business or logical constraints.

Statistical outliers are not automatically treated as data errors.
Extreme values will be evaluated based on their business context.

In [33]:
expected_ranges = {
    "Age": (18, 60),
    "DistanceFromHome": (1, None),
    "Education": (1, 5),
    "EnvironmentSatisfaction": (1, 4),
    "JobInvolvement": (1, 4),
    "JobLevel": (1, 5),
    "JobSatisfaction": (1, 4),
    "PerformanceRating": (3, 4),
    "RelationshipSatisfaction": (1, 4),
    "StockOptionLevel": (0, 3),
    "WorkLifeBalance": (1, 4),
    "TrainingTimesLastYear": (0, None),
    "NumCompaniesWorked": (0, None),
    "TotalWorkingYears": (0, None),
    "YearsAtCompany": (0, None),
    "YearsInCurrentRole": (0, None),
    "YearsSinceLastPromotion": (0, None),
    "YearsWithCurrManager": (0, None),
    "PercentSalaryHike": (0, None),
}

In [34]:
range_results = []

for column, (min_expected, max_expected) in expected_ranges.items():
    actual_min = df[column].min()
    actual_max = df[column].max()

    below_min = (
        actual_min < min_expected
        if min_expected is not None
        else False
    )

    above_max = (
        actual_max > max_expected
        if max_expected is not None
        else False
    )

    range_results.append({
        "column": column,
        "actual_min": actual_min,
        "actual_max": actual_max,
        "expected_min": min_expected,
        "expected_max": max_expected,
        "below_min": below_min,
        "above_max": above_max,
        "valid_range": not (below_min or above_max)
    })

range_results = pd.DataFrame(range_results)

range_results

,column,actual_min,actual_max,expected_min,expected_max,below_min,above_max,valid_range
0,Age,18,60,18,60.0,False,False,True
1,DistanceFromHome,1,29,1,NaN,False,False,True
2,Education,1,5,1,5.0,False,False,True
3,EnvironmentSatisfaction,1,4,1,4.0,False,False,True
4,JobInvolvement,1,4,1,4.0,False,False,True
5,JobLevel,1,5,1,5.0,False,False,True
6,JobSatisfaction,1,4,1,4.0,False,False,True
7,PerformanceRating,3,4,3,4.0,False,False,True
8,RelationshipSatisfaction,1,4,1,4.0,False,False,True
9,StockOptionLevel,0,3,0,3.0,False,False,True


In [35]:
numeric_columns = df.select_dtypes(include="number").columns

negative_values = {
    column: int((df[column] < 0).sum())
    for column in numeric_columns
}

negative_values = {
    column: count
    for column, count in negative_values.items()
    if count > 0
}

negative_values

{}

In [36]:
invalid_tenure = df[
    df["YearsAtCompany"] > df["TotalWorkingYears"]
]

len(invalid_tenure)

0

In [37]:
invalid_role_tenure = df[
    df["YearsInCurrentRole"] > df["YearsAtCompany"]
]

len(invalid_role_tenure)

0

In [38]:
invalid_manager_tenure = df[
    df["YearsWithCurrManager"] > df["YearsAtCompany"]
]

len(invalid_manager_tenure)

0

In [41]:
invalid_promotion = df[
    df["YearsSinceLastPromotion"] > df["YearsAtCompany"]
]

len(invalid_promotion)

0

In [43]:
outlier_summary = []

for column in numeric_columns:
    Q1 = df[column].quantile(0.25)
    Q3 = df[column].quantile(0.75)

    IQR = Q3 - Q1

    lower_bound = Q1 - 1.5 * IQR
    upper_bound = Q3 + 1.5 * IQR

    outlier_count = (
        (df[column] < lower_bound) |
        (df[column] > upper_bound)
    ).sum()

    outlier_summary.append({
        "column": column,
        "Q1": Q1,
        "Q3": Q3,
        "IQR": IQR,
        "lower_bound": lower_bound,
        "upper_bound": upper_bound,
        "outlier_count": int(outlier_count),
        "outlier_percentage": round(
            outlier_count / len(df) * 100,
            2
        )
    })

outlier_summary = pd.DataFrame(outlier_summary)

outlier_summary.sort_values(
    "outlier_count",
    ascending=False
)

,column,Q1,Q3,IQR,lower_bound,upper_bound,outlier_count,outlier_percentage
20,TrainingTimesLastYear,2.00,3.00,1.00,0.500,4.500,238,16.19
15,PerformanceRating,3.00,3.00,0.00,3.000,3.000,226,15.37
11,MonthlyIncome,2911.00,8379.00,5468.00,-5291.000,16581.000,114,7.76
24,YearsSinceLastPromotion,0.00,3.00,3.00,-4.500,7.500,107,7.28
22,YearsAtCompany,3.00,9.00,6.00,-6.000,18.000,104,7.07
18,StockOptionLevel,0.00,1.00,1.00,-1.500,2.500,85,5.78
19,TotalWorkingYears,6.00,15.00,9.00,-7.500,28.500,63,4.29
13,NumCompaniesWorked,1.00,4.00,3.00,-3.500,8.500,52,3.54
23,YearsInCurrentRole,2.00,7.00,5.00,-5.500,14.500,21,1.43
25,YearsWithCurrManager,2.00,7.00,5.00,-5.500,14.500,14,0.95


## Outlier Assessment Conclusion

Statistical outlier detection using the IQR method identified several
extreme observations across variables such as `TrainingTimesLastYear`,
`PerformanceRating`, `MonthlyIncome`, and career tenure variables.

However, these observations were not automatically removed.

Business-rule validation confirmed that the identified extreme values
fall within valid business ranges and do not violate logical relationships
between workforce variables.

For example:

- `PerformanceRating = 4` is a valid rating category even though it is
  statistically flagged because the IQR is zero.
- Higher `MonthlyIncome` values can be valid for employees with higher
  job levels.
- Higher tenure values represent valid long-term employees.
- Higher `TrainingTimesLastYear` values represent valid training activity.

Therefore, no observations will be removed solely based on statistical
outlier detection.

Extreme values will be retained for analysis unless a specific business
or data-quality rule identifies them as invalid.